### ISSUE TITLE
interpolation of file() should not fail

### ISSUE THREAD

**Issue by akranga** [hashicorp/terraform#6460]
Interpolation function file() fails if file not found however it shouldn't. Here is an example

```
resource "null_resource" "gen_ssh_key" {
    provisioner "local-exec" {
        command = "ssh-keygen blablabla -f ${path.cwd}/id_rsa"
    }
}

resource "aws_iam_user_ssh_key" "user" {
    depends_on = ["null_resource.gen_ssh_key"]
    ...
    public_key = "${file("${path.cwd}/id_rsa.pub")}"
}
```

It seem terraform tries to compute state of the attribute before execution. However file does not exists which leads to failure. Will be happy for any sort of workaround

**Comment by jen20** [comment]
Hi @akranga! You're correct with this at the moment - there is no workaround strictly within Terraform. One common method people are using here is to wrap Terraform execution with a script that generates out a key if necessary and then passes in the file path as a variable. Would this work for you?

**Comment by msnelling** [comment]
I'm also having trouble with this. I'm trying to work around the AWS provider not returning the directory security group when creating a directory instance (#5750). So far the only way I can see to do this is by by using a `null_resource` with a `local-exec` provisioner to run the `aws ec2 describe-security-groups` command and outputting the result to file.
```
resource "null_resource" "directory_sg_file" {
  triggers {
    directory_ids = "${aws_directory_service_directory.ad.id}"
  }
  
  provisioner "local-exec" {
    command = "aws --output text ec2 describe-security-groups --filters Name=group-name,Values=${aws_directory_service_directory.ad.id}_controllers --query SecurityGroups[].[GroupId] >> directory_sg.txt"
  }
}
```
However when trying to read the file contents into a variable with the `file()` interpolation function I get the error `The system cannot find the file specified`

TL;DR Please either add a flag to the `file()` interpolation function or change the default behaviour. (Even better, fix issue #5750 ;-)

**Comment by kesor** [comment]
I used an ugly hack to get around the problem. It involves having a file before terraform is run, but without content, or with some irrelevant content. This makes terraform not fall on its face when planning, and it does actually work changing this file during the `apply` execution and then reading the value in the file from other resources.

This method can also be used for a workaround around #2727.

```
resource "null_resource" "amazon_output" {
  triggers {
    always = "${uuid()}"
  }
  provisioner "local-exec" {
    command = "aws ec2 describe-instances --max-results=5 | jq -r '.Reservations | .[0].Instances | .[].InstanceId' > ${path.module}/single_instance_id.txt"
  }
}

resource null_resource "print_out_variable_from_file" {
  triggers {
    always = "${uuid()}"
  }
  provisioner "local-exec" {
    command = "echo ${trimspace(file(${path.module}/single_instance_id.txt))}"
  }
}
```

start it using something like:
```
touch ec2-instances/single_instance_id.txt
terraform apply
```

This is horrible -- please fix `file()` and variables interpolation, the horrors I had to write in my terraform files to make some trivial output be used in resources is just nightmarish.

**Comment by apparentlymart** [comment]
I know that the example given in the original comment here was just that -- an example -- but I just wanted to point out that for the first example of generating SSH keys we have a first-class Terraform resource that can do this:

```hcl
resource "tls_private_key" "ssh" {
  algorithm = "RSA"
}

resource "aws_iam_user_ssh_key" "user" {
    public_key = "${tls_private_key.ssh.public_key_openssh}"
}
```

`file()` checks at plan time to try to help people who have made a mistake in specifying the filename, so they can catch it with a `terraform plan` rather than having Terraform fail partway though apply.

The `template_file` resource used to be a reasonable workaround for reading files in a more robust way (as long as the file doesn't contain template interpolation syntax) but that's now deprecated. It could be interesting to have a `local_file` data source for reading files... it's not as convenient as the `file()` function would be of course, but having it be a proper resource block allows us to describe its dependencies properly so we can have more control over at what point in the process the file is read, and thus we can get the best of both worlds:

```
# hypothetical example... not actually implemented yet!
resource "null_resource" "amazon_output" {
  triggers {
    always = "${uuid()}"
  }
  provisioner "local-exec" {
    command = "aws ec2 describe-instances --max-results=5 | jq -r '.Reservations | .[0].Instances | .[].InstanceId' > ${path.module}/single_instance_id.txt"
  }
}

data "local_file" "single_instance_id" {
  filename = "${path.module}/single_instance_id.txt"

  # Depending on a resource means this won't be evaluated until
  # apply time, but if it didn't then we'd evaluate it at plan time.
  depends_on = ["null_resource.amazon_output"]
}

resource "null_resource" "print_out_variable_from_file" {
  triggers {
    always = "${uuid()}"
  }
  provisioner "local-exec" {
    command = "echo '${local_file.single_instance_id.contents}'"
  }
}
```

...though for all of these use-cases of running the `aws` CLI to retrieve data we could instead implement data sources to allow this data to be retrieved directly inside Terraform, similar to what I've been doing in #6819. As far as I'm concerned, if we can possibly avoid generating local files from provisioners then we should, since it'll always be more robust to make Terraform aware of what's going on.

**Comment by kesor** [comment]
@apparentlymart you might be right about implementing various provisioners as Terraform code, but this is also a usability issue. Even though the result is not pretty, it is much more accessible to people to use an output of some command as a variable vs. writing Go code (and in most cases required to learn Go first).

**Comment by apparentlymart** [comment]
@kesor certainly yes... I wasn't meaning to suggest that people should always write Go code if they have a problem to solve, but more that we should aim to have already solved the most interesting cases by taking these examples.

Another thing we could do, which didn't occur to me the first time, is to have a data source explicitly intended for running a local program and capturing its output to expose as an attribute, so we can support this sort of pattern without the need to create intermediate files on local disk.

**Comment by kesor** [comment]
@apparentlymart although offtopic, but to continue your brainstorming. The problem I was trying to solve involved creating a resource using AWS-CLI that is not yet supported in Terraform (AWS Cognito).

I guess it is a suggestion for another issue, but having the ability to "create" the resource and get its ID output is great. Having to then enable a *different* command that will be invoked during "destroy" and/or "update" is even more powerful. At the moment I created a shell script that does create-or-destroy and it breaks the whole idempotency thing, can share the code, its ugly.

**Comment by mitchellh** [comment]
I think changing the interpolation behavior during plan is too difficult. I think the correct solution here is going to be being able to assign "variables" during runtime, and then threading that through. By doing that, we'll be able to just turn the file load into a computed without any heavy interpolation changes. 

There is another issue tracking setting vars so I'll close this.
